HPCC-8835 Generate shared global classes instead of member classes

shamser opened the pull request:
Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

shamser committed:
HPCC-8835 Generate shared global classes instead of member classes

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

shamser commented:
@ghalliday Please can you review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 13c6d2811af3df7f66452c2c127f26f10526e446
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.224 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |769 |769 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on ecl/hqlcpp/hqlhtcpp.cpp:
```diff
@@ -5644,22 +5650,57 @@ void HqlCppTranslator::buildHashClass(BuildCtx & ctx, const char * name, IHqlExp
+            return;
+        }
+
+        gcc.createGlobalCompareInstance(buildctx, name);
```
Note compare functions are not being commoned up because of this - see aggsqx3, sortrev for example

ghalliday reviewed (changes requested):
@shamser some comments about structure, and a bug

shamser commented on ecl/hqlcpp/hqlhtcpp.ipp:
```diff
@@ -58,6 +58,48 @@ private:
     bool grouped;
 };
 
+class GlobalCompareClass
```
@ghalliday couldn't find GlobalClass.  Do you mean GlobalClassBuilder?

ghalliday replied:
@shamser I wasn't referring to an existing class - more musing about how I thought it would be structured.

Ideally it would also be combined with the GlobalClassBuilder, but that can wait for another day - these sort classes do not have access to ICodeContext, whereas classes created with the GlobalClassBuilder do.  Using a similar pattern for creating the class would make sense.

ghalliday replied:
So the GlobalClassBuilder doesn't really have a class that represents the class being created it is one that knows how to build a class like that.

shamser committed:
HPCC-8835 WIP

Signed-off-by: Shamser Ahmed <shamser.ahmed@lexisnexis.co.uk>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 82125cf8301d9824381727d689c033ba30d9838a
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.104 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |769 |769 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: 7b6e2758f505144edb373bbdd8a91c28352004d9
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 45.501 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |9 |2 |
|setup (thor) |11 |9 |2 |
|setup (roxie) |11 |9 |2 |
|test (hthor) |741 |667 |74 |
|test (thor) |636 |611 |25 |
|test (roxie) |769 |698 |71 |


[Errors:](https://gist.github.com/56bd0560c24643c75193c286d3434a37)
- test
  - hthor
    - 136\. [dedupwithbest](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-dedupwithbest-w20170710-113832-2-log), 159\. [dict15 (multiPart=false)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-dict15-multipart-false-w20170710-113836-9-log), 160\. [dict15 (multiPart=true)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-dict15-multipart-true-w20170710-113837-log), 161\. [dict15a (multiPart=false)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-dict15a-multipart-false-w20170710-113837-3-log), 162\. [dict15a (multiPart=true)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-dict15a-multipart-true-w20170710-113837-4-log), 163\. [dict15b (multiPart=false)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-dict15b-multipart-false-w20170710-113837-5-log), 164\. [dict15b (multiPart=true)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-dict15b-multipart-true-w20170710-113838-log), 165\. [dict15c (multiPart=false)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-dict15c-multipart-false-w20170710-113838-5-log), 166\. [dict15c (multiPart=true)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-dict15c-multipart-true-w20170710-113838-3-log), 248\. [globalmerge](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-globalmerge-w20170710-113858-5-log), 392\. [merge](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-merge-w20170710-113938-log), 393\. [mergechild (multiPart=false)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-mergechild-multipart-false-w20170710-113938-3-log), 394\. [mergechild (multiPart=true)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-mergechild-multipart-true-w20170710-113938-4-log), 395\. [mergededup](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-mergededup-w20170710-113938-6-log), 396\. [mergenway (multiPart=false)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-mergenway-multipart-false-w20170710-113938-5-log), 397\. [mergenway (multiPart=true)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-mergenway-multipart-true-w20170710-113938-7-log), 415\. [nwaytest](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-nwaytest-w20170710-113942-1-log), 457\. [prefetch (multiPart=true)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-prefetch-multipart-true-w20170710-113951-5-log), 458\. [prefetch2 (multiPart=true)](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-prefetch2-multipart-true-w20170710-113951-1-log), 599\. [stepjoin1](https://gist.github.com/HPCCSmoketest/56bd056
 ... ( comment is too long, 52 lines, 28990 bytes, truncated) ... 
://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-setupsearchindex-w20170710-113757-8-log), 9\. [setupwordindex](https://gist.github.com/HPCCSmoketest/56bd0560c24643c75193c286d3434a37#file-setupwordindex-w20170710-113757-10-log), 

```

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: bca0e26f6cfaae712517521bc167ec9d7d91d9b9
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 43.46 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |5 |6 |
|setup (thor) |11 |5 |6 |
|setup (roxie) |11 |5 |6 |
|test (hthor) |741 |343 |398 |
|test (thor) |636 |312 |324 |
|test (roxie) |769 |347 |422 |


[Errors:](https://gist.github.com/10de870f655697e16e4fa925604addff)
- test
  - hthor
    - 9\. [aggds1 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggds1-multipart-false,-usesequential-true-w20170710-123028-9-log), 8\. [aggds1 (multiPart=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggds1-multipart-true-w20170710-123028-2-log), 11\. [aggds2 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggds2-multipart-false,-usesequential-true-w20170710-123028-18-log), 10\. [aggds2 (multiPart=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggds2-multipart-true-w20170710-123028-20-log), 13\. [aggds3 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggds3-multipart-false,-usesequential-true-w20170710-123028-16-log), 12\. [aggds3 (multiPart=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggds3-multipart-true-w20170710-123028-6-log), 15\. [aggds4 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggds4-multipart-false,-usesequential-true-w20170710-123028-8-log), 14\. [aggds4 (multiPart=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggds4-multipart-true-w20170710-123028-3-log), 17\. [aggidx1 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggidx1-multipart-false,-usesequential-true-w20170710-123028-24-log), 16\. [aggidx1 (multiPart=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggidx1-multipart-true-w20170710-123028-7-log), 19\. [aggidx2 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggidx2-multipart-false,-usesequential-true-w20170710-123028-10-log), 18\. [aggidx2 (multiPart=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggidx2-multipart-true-w20170710-123028-12-log), 21\. [aggidx3 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggidx3-multipart-false,-usesequential-true-w20170710-123029-log), 20\. [aggidx3 (multiPart=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggidx3-multipart-true-w20170710-123028-13-log), 23\. [aggidx4 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggidx4-multipart-false,-usesequential-true-w20170710-123030-4-log), 22\. [aggidx4 (multiPart=true)](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggidx4-multipart-true-w20170710-123030-log), 25\. [aggregate](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-aggregate-w20170710-123030-3-
 ... ( comment is too long, 52 lines, 179051 bytes, truncated) ... 
sq](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-setupsq-w20170710-123017-8-log), 9\. [setupwordindex](https://gist.github.com/HPCCSmoketest/10de870f655697e16e4fa925604addff#file-setupwordindex-w20170710-123017-13-log), 

```

HPCCSmoketest commented:
Automated Smoketest: :x:
Sha: db4bed0a1b7059c21f61a0b15d4b028dceafae52
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.932 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |5 |6 |
|setup (thor) |11 |5 |6 |
|setup (roxie) |11 |5 |6 |
|test (hthor) |741 |343 |398 |
|test (thor) |636 |312 |324 |
|test (roxie) |769 |347 |422 |


[Errors:](https://gist.github.com/c40069706e4abb7ca9304eb2255e5901)
- test
  - hthor
    - 9\. [aggds1 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggds1-multipart-false,-usesequential-true-w20170710-141245-9-log), 8\. [aggds1 (multiPart=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggds1-multipart-true-w20170710-141245-7-log), 11\. [aggds2 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggds2-multipart-false,-usesequential-true-w20170710-141245-11-log), 10\. [aggds2 (multiPart=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggds2-multipart-true-w20170710-141245-13-log), 13\. [aggds3 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggds3-multipart-false,-usesequential-true-w20170710-141245-3-log), 12\. [aggds3 (multiPart=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggds3-multipart-true-w20170710-141245-17-log), 15\. [aggds4 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggds4-multipart-false,-usesequential-true-w20170710-141245-21-log), 14\. [aggds4 (multiPart=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggds4-multipart-true-w20170710-141245-16-log), 17\. [aggidx1 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggidx1-multipart-false,-usesequential-true-w20170710-141245-6-log), 16\. [aggidx1 (multiPart=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggidx1-multipart-true-w20170710-141245-4-log), 19\. [aggidx2 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggidx2-multipart-false,-usesequential-true-w20170710-141245-19-log), 18\. [aggidx2 (multiPart=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggidx2-multipart-true-w20170710-141245-10-log), 21\. [aggidx3 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggidx3-multipart-false,-usesequential-true-w20170710-141246-log), 20\. [aggidx3 (multiPart=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggidx3-multipart-true-w20170710-141245-8-log), 23\. [aggidx4 (multiPart=false, useSequential=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggidx4-multipart-false,-usesequential-true-w20170710-141246-3-log), 22\. [aggidx4 (multiPart=true)](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggidx4-multipart-true-w20170710-141246-4-log), 25\. [aggregate](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-aggregate-w20170710-141247
 ... ( comment is too long, 52 lines, 179062 bytes, truncated) ... 
sq](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-setupsq-w20170710-141233-11-log), 9\. [setupwordindex](https://gist.github.com/HPCCSmoketest/c40069706e4abb7ca9304eb2255e5901#file-setupwordindex-w20170710-141233-7-log), 

```

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: 49a6d63ddd9635f0c5ab52956bb1291a97663b49
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 43.774 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |769 |769 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

shamser commented:
@ghalliday I've removed the GlobalCompareclass as it was far simpler to embed the functionality withing buildCompareClass.  I've replaced the 3 lines to create class + helper function with the a single call to buildCompareFuncHelper.  ItThe helper function's name is derived from the CompareFuncName ("query"+compareFuncName).  Also, when global the object name is now all prefixed with "compare" as they may be shared between queryEqualCompare, queryMergeCompare, queryCompare etc.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
Sha: ec3a3032f660ecdbcd2c91b93bb4a8624f2f6b40
Build: success
Build: success
ECL Watch:  Rebuilding Site

| errors | warnings | build time | 
|---|---|---|
| 0 | 75 | 44.736 seconds |


Install hpccsystems-platform-community_6.5.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |741 |741 |0 |
|test (thor) |636 |636 |0 |
|test (roxie) |769 |769 |0 |


HPCC Stop: OK
HPCC Uninstall: OK

ghalliday commented on ecl/hqlcpp/hqlhtcpp.cpp:
```diff
@@ -5644,22 +5650,62 @@ void HqlCppTranslator::buildHashClass(BuildCtx & ctx, const char * name, IHqlExp
+    translator.buildReturnOrder(func.ctx, sortList, dataset);
+}
 
+void HqlCppTranslator::buildCompareClass(BuildCtx & ctx, const char * name, IHqlExpression * sortList, const DatasetReference & dataset, StringBuffer & compareFuncName)
```
picky: compareFuncName should really be compareObjectName

ghalliday commented:
That looks good.  I'll run some regression test changes, but assuming no problems I am happy to merge as-is
